internal/xds/resolver/xds_resolver.go

```diff
@@ -111,50 +111,45 @@ type xdsResolverBuilder struct {
 // The xds bootstrap process is performed (and a new xDS client is built) every
 // time an xds resolver is built.
 func (b *xdsResolverBuilder) Build(target resolver.Target, cc resolver.ClientConn, opts resolver.BuildOptions) (_ resolver.Resolver, retErr error) {
+	// Initialize the xDS client.
+	newXDSClient := rinternal.NewXDSClient.(func(string, estats.MetricsRecorder) (xdsclient.XDSClient, func(), error))
+	if b.newXDSClient != nil {
+		newXDSClient = b.newXDSClient
+	}
+	client, xdsClientClose, err := newXDSClient(target.String(), opts.MetricsRecorder)
+	if err != nil {
+		return nil, fmt.Errorf("xds: failed to create xds-client: %v", err)
+	}
+
+	template, err := sanityChecksOnBootstrapConfig(target, opts, client)
+	if err != nil {
+		xdsClientClose()
+		return nil, err
+	}
+	ldsResourceName := bootstrap.PopulateResourceTemplate(template, target.Endpoint())
+
 	r := &xdsResolver{
-		cc:             cc,
-		activeClusters: make(map[string]*clusterInfo),
-		channelID:      rand.Uint64(),
+		cc:              cc,
+		xdsClient:       client,
+		xdsClientClose:  xdsClientClose,
+		activeClusters:  make(map[string]*clusterInfo),
+		channelID:       rand.Uint64(),
+		ldsResourceName: ldsResourceName,
 	}
-	defer func() {
-		if retErr != nil {
-			r.Close()
-		}
-	}()
 	r.logger = prefixLogger(r)
 	r.logger.Infof("Creating resolver for target: %+v", target)
 
 	// Initialize the serializer used to synchronize the following:
-	// - updates from the xDS client. This could lead to generation of new
-	//   service config if resolution is complete.
+	// - updates from the dependency manager. This could lead to generation of
+	//   new service config if resolution is complete.
 	// - completion of an RPC to a removed cluster causing the associated ref
 	//   count to become zero, resulting in generation of new service config.
 	// - stopping of a config selector that results in generation of new service
 	//   config.
 	ctx, cancel := context.WithCancel(context.Background())
 	r.serializer = grpcsync.NewCallbackSerializer(ctx)
 	r.serializerCancel = cancel
-
-	// Initialize the xDS client.
-	newXDSClient := rinternal.NewXDSClient.(func(string, estats.MetricsRecorder) (xdsclient.XDSClient, func(), error))
-	if b.newXDSClient != nil {
-		newXDSClient = b.newXDSClient
-	}
-	client, closeFn, err := newXDSClient(target.String(), opts.MetricsRecorder)
-	if err != nil {
-		return nil, fmt.Errorf("xds: failed to create xds-client: %v", err)
-	}
-	r.xdsClient = client
-	r.xdsClientClose = closeFn
-
-	// Determine the listener resource name and start a watcher for it.
-	template, err := r.sanityChecksOnBootstrapConfig(target, opts, r.xdsClient)
-	if err != nil {
-		return nil, err
-	}
-	r.dataplaneAuthority = opts.Authority
-	r.ldsResourceName = bootstrap.PopulateResourceTemplate(template, target.Endpoint())
-	r.listenerWatcher = newListenerWatcher(r.ldsResourceName, r)
+	r.dm = xdsdepmgr.New(r.ldsResourceName, opts.Authority, r.xdsClient, r)
```

Comment:
Can we not make these three fields also be part of the literal struct initialization?